Source/Core/Core/HW/EXI_DeviceAGP.cpp

```diff
@@ -0,0 +1,287 @@
+// Copyright 2013 Dolphin Emulator Project
+// Licensed under GPLv2
+// Refer to the license.txt file included.
+
+#include "Common/FileUtil.h"
+#include "Common/MemoryUtil.h"
+#include "Common/StdMakeUnique.h"
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/HW/EXI_Device.h"
+#include "Core/HW/EXI_DeviceAGP.h"
+#include "Core/HW/Memmap.h"
+
+CEXIAgp::CEXIAgp(int index)
+{
+	m_slot = index;
+
+	// Create the ROM
+	m_rom_size = 0;
+
+	LoadRom();
+
+	m_address = 0;
+	m_rom_hash_loaded = false;
+}
+
+CEXIAgp::~CEXIAgp()
+{
+	std::string path;
+	std::string filename;
+	std::string ext;
+	std::string gbapath;
+	SplitPath(m_slot == 0 ? SConfig::GetInstance().m_strGbaCartA : SConfig::GetInstance().m_strGbaCartB, &path, &filename, &ext);
+	gbapath = path + filename;
+
+	SaveFileFromEEPROM(gbapath + ".sav");
+}
+
+void CEXIAgp::DoHash(u8* data, u32 size)
+{
+	if (!m_rom_hash_loaded)
+		LoadHash();
+
+	for (u32 it = 0; it < size; it++)
+	{
+		m_hash = m_hash ^ data[it];
+		m_hash = m_hash_array[m_hash];
+	}
+}
+
+void CEXIAgp::LoadRom()
+{
+	// Load whole ROM dump
+	std::string path;
+	std::string filename;
+	std::string ext;
+	std::string gbapath;
+	SplitPath(m_slot == 0 ? SConfig::GetInstance().m_strGbaCartA : SConfig::GetInstance().m_strGbaCartB, &path, &filename, &ext);
+	gbapath = path + filename;
+	LoadFileToROM(gbapath + ext);
+	INFO_LOG(EXPANSIONINTERFACE, "Loaded GBA rom: %s card: %d", gbapath.c_str(), m_slot);
+	LoadFileToEEPROM(gbapath + ".sav");
+	INFO_LOG(EXPANSIONINTERFACE, "Loaded GBA sav: %s card: %d", gbapath.c_str(), m_slot);
+}
+
+void CEXIAgp::LoadFileToROM(std::string filename)
+{
+	File::IOFile pStream(filename, "rb");
+	if (pStream)
+	{
+		u64 filesize = pStream.GetSize();
+		m_rom_size = filesize & 0xFFFFFFFF;
+		m_rom_mask = (m_rom_size - 1);
+
+		m_rom.resize(m_rom_size);
+
+		pStream.ReadBytes(m_rom.data(), filesize);
+	}
+	else
+	{
+		// dummy rom data
+		m_rom.resize(0x2000);
+	}
+}
+
+void CEXIAgp::LoadFileToEEPROM(std::string filename)
+{
+	File::IOFile pStream(filename, "rb");
+	if (pStream)
+	{
+		u64 filesize = pStream.GetSize();
+		m_eeprom_size = filesize & 0xFFFFFFFF;
+		m_eeprom_mask = (m_eeprom_size - 1);
+
+		m_eeprom.resize(m_eeprom_size);
+
+		pStream.ReadBytes(m_eeprom.data(), filesize);
+	}
+}
+
+void CEXIAgp::SaveFileFromEEPROM(std::string filename)
+{
+	File::IOFile pStream(filename, "wb");
+	if (pStream)
+	{
+		pStream.WriteBytes(m_eeprom.data(), m_eeprom_size);
+	}
+}
+
+void CEXIAgp::LoadHash()
+{
+	if (!m_rom_hash_loaded && m_rom_size > 0)
+	{
+		for (int i = 0; i < 0x100; i++)
+		{
+			// Load the ROM hash expected by the AGP
+			m_hash_array[i] = Memory::ReadUnchecked_U8(0x0017e908 + i);
```

Comment:
I just noticed this line; why is the behavior of the AGP device dependent on the contents of an arbitrary address in memory?  (It might be that we don't have any choice... but at the very least, it needs a better comment.